microsoft/PowerToys
# [Keyboard Manager] Add JSON support for App Specific shortcuts

## Summary of the Pull Request
This PR allows app specific shortcuts to be saved to JSON, so that it can be reloaded on re-opening PowerToys and so that the changes are reflected on the Landing page. The landing page shortcut list has also been modified to include the target application.

This PR depends on #4804 
Diff between this PR and #4804
https://github.com/arjunbalgovind/PowerToys/compare/user/arbalgov/appSpecificUI...arjunbalgovind:user/arbalgov/appSpecificJSON
![image](https://user-images.githubusercontent.com/32061677/86860240-53ef7800-c079-11ea-9705-9e002af772a7.png)


## PR Checklist
* [X] Applies to #4713 


## Detailed Description of the Pull Request / Additional comments
- Changed the JSON format to have an additional field under `remapShortcuts` called `appSpecific`, which in turn has a similar array structure as the global shortcuts array, but with an additional field `targetApp`. The following is an example:
`{"remapKeys":{"inProcess":[]},"remapShortcuts":{"global":[{"originalKeys":"162;67","newRemapKeys":"162;86"},{"originalKeys":"162;68","newRemapKeys":"162;86"}],
"appSpecific":[{"originalKeys":"162;67","newRemapKeys":"162;86","targetApp":"msedge"},{"originalKeys":"162;68","newRemapKeys":"162;86","targetApp":"chrome"}]
}}`
When we combine both the remap keys and shortcuts windows since we will need upgrade path logic at that point, we can refactor the JSON structure such that there is only one remap section and all the elements have a `targetApp` element.
- While loading the JSON in the backend, separate try catch blocks have been added for each section, so that even if Single Key Remaps fail to load, OS level shortcuts can load, etc. This allows the upgrade path logic to work since an exception will occur while loading app specific shortcuts but the other remaps will not be affected.
- On the PT Settings side, a new data model `AppSpecificKeysDataModel` was added which inherits from the old `KeysDataModel`, with an additional field for `TargetApp`. The `GetOriginalKeys` and `GetNewRemapKeys` methods were implemented which returned the base method because there were warnings on the UI Xaml side which said the method did not exist of that data model. The `appSpecific` JSON property was linked to this data model.
- The Remap shortcuts ListView was modified such that it has an additional arrow followed by the target app name
- The ListView itemSource was linked to a combination of the os level remap list and the app specific remap list, where the os level remaps have the target app name "All Apps".
- Since the data template used for Remap Keys could no longer be reused for Remap Shortcuts, a separate data template was made for this listview.
- Horizontal scrolling was added for the listview since with the additional target app column it could easily overflow.

## Known Issues
- After saving the settings, the order in which shortcuts appear will not be the same in the landing page and on re-opening the Remap Shortcuts UI because it loads the OS level shortcuts first and then the app specific shortcuts.
- Columns were not aligned in the list view since KBM was released, however it is more noticeable now.


## Validation Steps Performed
- Opened Remap Shortcuts and the Keyboard Manager landing page with the old JSON file to ensure no exceptions occur in the upgrade path.
- Added multiple shortcuts with and without target apps, and checked if the landing page was updated and if re-opening Remap shortcuts shows the correct remaps
- Closed PT and re-opened to ensure all the saved mappings are loaded
- Added three key shortcut remaps with long app names to check if horizontal scrolling is working
